servicemanager/servicemanager_test.go

```diff
@@ -0,0 +1,545 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// Copyright (C) 2022 Renesas Electronics Corporation.
+// Copyright (C) 2022 EPAM Systems, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package servicemanager_test
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/aoscloud/aos_common/aoserrors"
+	"github.com/aoscloud/aos_common/image"
+	"github.com/opencontainers/go-digest"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/aoscloud/aos_servicemanager/config"
+	"github.com/aoscloud/aos_servicemanager/servicemanager"
+)
+
+func init() {
+	log.SetFormatter(&log.TextFormatter{
+		DisableTimestamp: false,
+		TimestampFormat:  "2006-01-02 15:04:05.000",
+		FullTimestamp:    true,
+	})
+	log.SetLevel(log.DebugLevel)
+	log.SetOutput(os.Stdout)
+}
+
+/**********************************************************************************************************************
+ * Consts
+ *********************************************************************************************************************/
+
+const (
+	errorAddServiceID = "errorAddServiceID"
```

Comment:
Add `err error` field to `testServiceInfoProvider` and set it when you would like to return error.

Reply:
this is not an error, it is ID  to simulate error case

Reply:
I understand but it is not obvious that using special ID causes the specific error. I suggest another approach:
```go
 type testServiceStorage struct {
    Services []servicemanager.ServiceInfo
    err Errror   
}

func (storage *testServiceStorage) GetService(serviceID string) (service servicemanager.ServiceInfo, err error) {
    if storage.err != nil {
        return storage.err
    }
    ...
}
```
In this case you have more control which error and when to return.
```go
storage.err = servicemanager.ErrNotExitst

if err := storage.GetService(serviceID); err != nil {
    ...
}
```